lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -802,6 +803,263 @@ func::FuncOp PolynomialToStandard::buildPolynomialModFunc(FunctionType funcType,
+        rootsType,
+        DenseElementsAttr::get(rootsType, precomputeRoots(root, cmod, degree)));
+
+    // Here is a slightly modified implementation of the standard iterative NTT
+    // computation using Cooley-Turkey butterfly. For reader reference:
+    // https://doi.org/10.1007/978-3-031-46077-7_22, and,
+    // https://doi.org/10.1109/ACCESS.2023.3294446
+    //
+    // We modify the standard implementation by pre-computing the root
+    // exponential values during compilation instead of doing so at runtime.
+    //
+    // Let roots be a tensor of <n x ix> where roots[i] = \psi^i, n be the
+    // degree of the polynomial and a be the coefficients. Then we implement the
+    // following:
+    //
+    // def fastNTT(a, n, cmod, roots):
+    //  a = BitReverseOrder(a)
+    //  m = n / 2
+    //  for (s = 0; s < log2(n); s++):
+    //    for (k = 0; k < n / m; k++):
+    //      for (j = 0; j < m / 2; j++):
+    //        A = a[k + j]
+    //        B = roots[(2 * j + 1) * n / m] * a[k + j + m / 2]
+    //        a[k + j] = A + B mod cmod
+    //        a[k + j + m / 2] = A - B mod cmod
+    //      end
+    //    end
+    //    m = m * 2
+    //  end
+
+    // Convert the input into Reverse-Bit Order
+    initialValue = computeReverseBitOrder(b, intermediateType, initialValue);
+
+    // Initialize the variables
+    Value initialBatchSize = b.create<arith::ConstantOp>(
+        IndexType::get(b.getContext()), b.getIndexAttr(2));
+
+    Value zero = b.create<arith::ConstantOp>(IndexType::get(b.getContext()),
+                                             b.getIndexAttr(0));
+    Value two = b.create<arith::ConstantOp>(IndexType::get(b.getContext()),
+                                            b.getIndexAttr(2));
+
+    Value n = b.create<arith::ConstantOp>(IndexType::get(b.getContext()),
+                                          b.getIndexAttr(degree));
+
+    // Define index affine mappings
+    AffineExpr x, y;
+    bindDims(b.getContext(), x, y);
+
+    auto nttIP = b.saveInsertionPoint();
+    auto stagesLoop = b.create<affine::AffineForOp>(
+        /*lowerBound=*/0, /* upperBound=*/stages, /*step=*/1,
+        /*iterArgs=*/ValueRange{initialValue, initialBatchSize});
+    b.setInsertionPointToStart(stagesLoop.getBody());
+    {
+      Value target = stagesLoop.getRegionIterArgs()[0];
+      Value batchSize = stagesLoop.getRegionIterArgs()[1];
+
+      // Compute this here as opposed to in the innerLoop as it is used later
+      Value rootExp = b.create<affine::AffineApplyOp>(x.floorDiv(y),
+                                                      ValueRange{n, batchSize});
+
+      auto stagesIP = b.saveInsertionPoint();
+      auto innerLoop = b.create<affine::AffineForOp>(
+          /*lbOperands=*/zero, /*lbMap=*/AffineMap::get(1, 0, x),
+          /*ubOperands=*/rootExp, /*ubMap=*/AffineMap::get(1, 0, x),
+          /*step=*/1, /*iterArgs=*/target);
+      b.setInsertionPointToStart(innerLoop.getBody());
+      {
```

Comment:
You can avoid the insertion point management and manual nesting by using the `BodyBuilderFn` builder, see for example

https://github.com/llvm/llvm-project/blob/7b06ec073d0bd8abad1274d12900a849bd3829c5/mlir/lib/Dialect/Affine/Transforms/SuperVectorize.cpp#L1357-L1360

You should just be able to give the last argument of `create` as

```
[&](OpBuilder &, Location, Value, ValueRange) {
  ...
}
```

And replace `...` with all the code below.